encode.go

```diff
@@ -10,107 +10,77 @@ import (
-		for i := 0; i < len(toSend); i++ {
-			err = writeData(dbuf, order, toSend[i])
+		for _, toSend := range data.Data.([]*K) {
+			err := writeData(dbuf, order, toSend)
 			if err != nil {
 				return err
 			}
 		}
-	// Scalar types backed by any variant of boolean, int, float, bytes
-	// * Reminder: UUID is back by byte array of length 16
-	case -KB, -UU, -KG, -KH, -KI, -KJ, -KE, -KF, -KC: // TODO: case: -KM , -KD, -KZ, -KN, -KU, -KV, -KT
-		binary.Write(dbuf, order, data.Type)
+	case -KB, -UU, -KG, -KH, -KI, -KJ, -KE, -KF, -KC, -KD, -KT, // TODO: case: -KM , -KZ, -KN, -KU, -KV
+		KB, UU, KG, KH, KI, KJ, KE, KF, KM, KD, KZ, KN, KU, KV, KT:
+		// Scalar & Vector of boolean, int variants, float variant, and bytes.
+		// Note: UUID is back by byte array of length 16
 		binary.Write(dbuf, order, data.Data)
-	// Vector types backed by any variant of boolean, int, float, bytes
-	case KB, UU, KG, KH, KI, KJ, KE, KF, KM, KD, KZ, KN, KU, KV, KT:
-		binary.Write(dbuf, order, data.Type)
-		binary.Write(dbuf, order, data.Attr)
-		binary.Write(dbuf, order, int32(reflect.ValueOf(data.Data).Len()))
-		binary.Write(dbuf, order, data.Data)
-	// String
-	case KC:
-		toSend := data.Data.(string)
-
-		binary.Write(dbuf, order, data.Type)
-		binary.Write(dbuf, order, data.Attr)
-		binary.Write(dbuf, order, int32(len(toSend)))
-		binary.Write(dbuf, order, []byte(toSend))
-	// Scalar symbol
-	case -KS:
-		toSend := data.Data.(string)
-
-		binary.Write(dbuf, order, data.Type)
-		binary.Write(dbuf, order, []byte(toSend))
-		binary.Write(dbuf, order, byte(0))
-	// Vector symbol
-	case KS:
-		toSend := data.Data.([]string)
-
-		binary.Write(dbuf, order, data.Type)
-		binary.Write(dbuf, order, data.Attr)
-		binary.Write(dbuf, order, int32(len(toSend)))
-		for i := 0; i < len(toSend); i++ {
-			binary.Write(dbuf, order, []byte(toSend[i]))
-			binary.Write(dbuf, order, byte(0))
+	case KC: // String
+		dbuf.WriteString(data.Data.(string))
+	case -KS: // Scalar symbol
+		dbuf.WriteString(data.Data.(string))
+		binary.Write(dbuf, order, byte(0)) // Null terminator
+	case KS: // Vector symbol
+		for _, toSend := range data.Data.([]string) {
+			dbuf.WriteString(toSend)
+			binary.Write(dbuf, order, byte(0)) // Null terminator
 		}
-	case -KP:
-		toSend := data.Data.(time.Time)
-
-		binary.Write(dbuf, order, data.Type)
-		binary.Write(dbuf, order, toSend.Sub(qEpoch))
-	case KP:
-		binary.Write(dbuf, order, data.Type)
-		binary.Write(dbuf, order, data.Attr)
-		binary.Write(dbuf, order, int32(reflect.ValueOf(data.Data).Len()))
-		tosend := data.Data.([]time.Time)
-		for _, ts := range tosend {
-			binary.Write(dbuf, order, ts.Sub(qEpoch))
+	case -KP: // Scalar timestamp
+		binary.Write(dbuf, order, data.Data.(time.Time).Sub(qEpoch))
+	case KP: // Vector timestamp
+		for _, toSend := range data.Data.([]time.Time) {
+			binary.Write(dbuf, order, toSend.Sub(qEpoch))
 		}
-	case XD:
-		tosend := data.Data.(Dict)
-		binary.Write(dbuf, order, XD)
```

Comment:
What was this write() call doing before?

Reply:
It was writing the type (`XD`), which we now do at the start of the function for every type